### PR TITLE
Issue 7349 : Expose segment reader.

### LINKED ISSUE #7349: Expose segment reader as a public API.
**Is your feature request related to a problem? Please describe.**
Expose segment reader as a public API. It will allow user to read a specific segment. The API should be able to serialize ( implementation of from_bytes and to_bytes).

**Describe the solution you'd like**
Expose segment reader API from factory.

**Describe alternatives you've considered**
N/A

**Additional context**
Add unit test and integration test for newly added API.

### PR DESCRIPTION
**Pull request by Bhupender-Y**
**Change log description**  
Expose segment reader as a public API. 

**Purpose of the change**  
Fixes #7349 

**What the code does**  
Expose segment reader as a public API. 

**How to verify it**  
Unit test and integration test having this functionality should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/stream/SegmentReader.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Gets the next event in the segment. If there are no events currently available this will block up for
+     * timeoutMillis waiting for them to arrive. If none do, an EventReadWithStatus {@link EventReadWithStatus} will be returned with null for
+     * {@link EventReadWithStatus#getEvent()}.
+     *
+     * @param timeoutMillis An upper bound on how long the call may block before returning null.
+     * @return An instance of {@link EventReadWithStatus}, which contains the next event in the segment. In the case the timeoutMillis
+     *         is reached, {@link EventReadWithStatus#getEvent()} returns null.
+     */
+    EventReadWithStatus<T> read(long timeoutMillis);
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/SegmentReader.java:33]: Take a look at how https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface#FLIP27:RefactorSourceInterface-SourceReaderSourceReader  `pollNext` and `isAvailable` are implemented. The read method itself is non-blocking but sync, where as isAvailable returns a future. 

We can implement this by having the impl use a SegmentInputStream's read with a buffer no larger than the the value returned via `bytesInBuffer` and use it's `fillBuffer` method to request more data.

File: client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            return read != null ? deserializer.deserialize(read) : null;
+        } catch (NoSuchSegmentException | SegmentTruncatedException e) {
+            handleSegmentTruncated();
+            throw new TruncatedDataException("Segment " + segment + " has been truncated.");
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return input.fillBuffer().thenCompose(x -> CompletableFuture.completedFuture(null));
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java:78]: Futures.toVoid

**Review by tkaitchuck — APPROVED** [review]

File: client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java (hunk 2 above)
Comment (Bhupender-Y) [reply client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java]: Done

File: client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return Futures.toVoid(input.fillBuffer());
+    }
+
+    private void handleSegmentTruncated()  {
+        log.info("{} encountered truncation for segment while read{} ", this, segment);
+        try {
+            input.setOffset(Futures.getThrowingExceptionWithTimeout(metadataClient.fetchCurrentSegmentHeadOffset(),
+                    clientConfig.getConnectTimeoutMilliSec()));
```
Comment (AnkurAnand11) [inline client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java:85]: If we encounter NoSuchSegmentException in the caller of handleSegmentTruncated(), what is the expectation here ? what offset are we setting here ?

**Review by AnkurAnand11 — APPROVED** [review]

**Review by anju-c-das — APPROVED** [review]
LGTM

**Review by ShwethaSNayak — APPROVED** [review]
LGTM

**Pull request merged**